src/libPMacc/include/particles/IdProvider.hpp

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright 2016 Alexander Grund
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "Environment.hpp"
+#include "eventSystem/EventSystem.hpp"
+#include "algorithms/reverseBits.hpp"
+#include "nvidia/atomic.hpp"
+#include "memory/buffers/HostDeviceBuffer.hpp"
+
+namespace PMacc {
+
+    namespace IdDetail {
+
+        __device__ uint64_cu nextId;
+
+        __global__ void setNextId(uint64_cu id)
+        {
+            nextId = id;
+        }
+
+        template<class T_Box>
+        __global__ void getNextId(T_Box boxOut)
+        {
+            boxOut(0) = nextId;
```

Comment:
For a DataBox this should result in a compile time error because there should be no cast operator from `int` to `DataBox<DIM1>`